src/sumo_gym/envs/fmp.py

```diff
@@ -79,58 +79,115 @@ def __init__(
+                old_edge_end_idx = self.edges[self.edge_dict[edge_id]].end
+
+                new_edge1 = Edge(old_edge_start_idx, vtx_counter)
+                new_edge2 = Edge(vtx_counter, old_edge_end_idx)
+                self.edges.append(new_edge1)
+                self.edges.append(new_edge2)
+
+                # instantiate new ChargingStation with location set to vtx_counter
+                self.charging_stations.append(ChargingStation(vtx_counter, 220, charging_station[3]))
+                
+                vtx_counter += 1
                 counter += 1
 
-                departures.append(self.edges[self.edge_dict[vehicle[1]]].start)
+            self.charging_stations = np.array(self.charging_stations)
+            self.vertices = np.array(self.vertices)
 
-            self.electric_vehicles = np.asarray(electric_vehicles)
-            self.departures = np.asarray(departures)
+            # remove indices from self.edges and fix self.edge_dict
+            # sort the indices in reverse order and pop the indices
+            # since the only elements in the list with different indices
+            # after a pop are those that reside on indices after the popped index
+            edge_idx_delete = sorted(list(set(edge_idx_delete)), reverse=True)
+            for idx in edge_idx_delete:
+                self.edges.pop(idx)
+            self.edges = np.array(self.edges)
 
-            demand = []
-            for d in raw_demand:
-                demand.append(Demand(self.vertex_dict[d[0]], self.vertex_dict[d[1]]))
+            # now, we need to repopulate self.edge_dict since the indices have changed
+            self.edge_dict = {} # sumo edge_id to idx in self.edges
+            counter = 0
+            for edge in self.edges:
+                if edge in self.revert_edge_dict:
+                    sumo_edge_id = self.revert_edge_dict[edge]
+                    self.edge_dict[sumo_edge_id] = counter
+                counter += 1
 
-            self.demand = np.asarray(demand)
+            self.electric_vehicles = []
+            self.ev_dict = {} # ev id to idx in self.electric_vehicles
+            counter = 0
+            for vehicle in raw_electric_vehicles:
+                self.electric_vehicles.append(ElectricVehicles(counter, vehicle[1], 220, vehicle[2]))
+                self.ev_dict[vehicle[0]] = counter
+                counter += 1
+            self.electric_vehicles = np.array(self.electric_vehicles)
+
+            # departure should be defined for all vehicles, so
+            # self.departures[i] should be the edge idx in self.edges of self.electric_vehicles[i]
+            self.departures = np.zeros(len(self.electric_vehicles))
+            for dpt in raw_departures:
+                self.departures[self.ev_dict[dpt[0]]] = self.edges[self.edge_dict[dpt[1]]].start
+            self.departures = np.array(self.departures)
+            self.departures = [int(num) for num in self.departures]
```

Comment:
please align with the variable typing, we just need a vertex index, to resolve the "starting edge" issue, you can assign the "edge_to" as the starting index, since the vehicle must finish traveling along the starting edge.

BTW, can you please check the turnaround issue I mentioned in the Monday meeting?

Reply:
I was under the impression that we only cared about the starting vertex of the starting edge provided in the sumo file for departures (in tutorial/make-fmpenv-data-simple.py)

Reply:
what was the issue again I kind of forgot

Reply:
> 

this is an issue I talked about in Monday's meeting, we need to consider how we are going to simulate the action consequence in later-on simulation. If you pass in just the start index, for example, (A -> B) as the start edge in sumo, but we route it to (A -> C) in model, it will raise conflict when calling traCI.

So:
1. use end index for departure so that each vehicle can at least finish its starting edge travel
2. turnaround issue: if we route A -> B, can we route B -> immediately, i.e., are we able to make turnaround at every junction in sumo.

Reply:
wait but how would the env know where to start if we only keep the end vertex. should we do edge idx instead?

Reply:
I think so but let me check

Reply:
> 

My thought is that you will help us to extract the "from" and "to" information, and then we will use "to" as the starting point for model sample func, but for sumo simulation, we will first let each vehicle arrive the "to" vertex from "from" vertex.